tools/foreman-cockpit-session

```diff
@@ -0,0 +1,248 @@
+#!/usr/bin/env ruby
+
+require "json"
+require "net/http"
+require "net/ssh"
+
+def dbg(str)
+  # $stderr.write(str + "\n")
+end
+
+# Cockpit protocol, encoding and decoding of control messages.
+
+def send_control(msg)
+  text = JSON.dump(msg)
+  dbg("-> #{text}")
+  $stdout.write("#{text.length+1}\n\n#{text}")
+  $stdout.flush
+end
+
+def read_size(io)
+  size = 0
+  seen = 0
+
+  while seen < 8
+    t = io.readpartial(1)
+
+    unless t
+      return 0
+    end
+
+    if t == "\n"
+      break
+    end
+
+    if t.to_i.zero? && t != "0"
+      raise ArgumentError("Invalid frame: invalid size")
+    end
+
+    size = (size * 10) + t.to_i
+    seen += 1
+  end
+
+  if seen == 8
+    raise ArgumentError("Invalid frame: size too long")
+  end
+
+  size
+end
+
+def read_control()
+  io = IO.for_fd(1)
+  size = read_size(io)
+
+  data = ""
+  while size > 0
+    d = io.readpartial(1)  # XXX - can't we just do readpartial(size)?
```

Comment:
Not sure why we can't just `io.read(size)`. We know that a message with that size is coming. And even if it doesn't, we can't really do anything with an incomplete message

Reply:
Change do use `$stdin.read(size)` and it works just fine.